glfs.c

```diff
@@ -263,6 +267,38 @@ static void gluster_cache_refresh(glfs_t *fs, const char *cfgstring)
 	}
 }
 
+static int gluster_cache_query_or_add(glfs_t **fs, gluster_server *entry,
+                                      char *config, bool *init)
+{
+	int ret = -1;
+
+	pthread_mutex_lock(&glfs_lock);
+
+	*fs = gluster_cache_query(entry, config);
+	if (*fs) {
+		*init = false;
+		ret = 0;
+		goto out;
+	}
+
+	*fs = glfs_new(entry->volname);
+	if (!*fs) {
+		tcmu_err("glfs_new failed: %m\n");
```

Comment:
Anything that you call that has a thread cancellation point needs to be taken into account when you lock a mutex.  http://pubs.opengroup.org/onlinepubs/009695399/functions/xsh_chap02_09.html  Search in that page for Cancellation Points.  That includes most forms of I/O.  In this case, if this thread was cancelled it could hit a cancellation point here and never unlock the mutex.  Search for uses of pthread_cleanup_push in tcmur_aio.c for example.

Also, isn't tcmu_dev_err more appropriate now?